### PR TITLE
[Fix][Connector kafka]Fix Kafka consumer stop fetching after TM node restarted

### LINKED ISSUE #7230: [Bug] [Connector Kafka] Kafka consumer stop fetching after TM node restarted
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/seatunnel/issues?q=is%3Aissue+label%3A%22bug%22) and found no similar issues.

### What happened

I'm run a seatunnel job with the Kafka connector on the Flink engine on the remote flink cluster,  after the TM encountered a runtime exception, that job entered the restart mechanism(fixed-day 1s). After Flink's restart mechanism restored that job, we found the Kafka connector stopped ingesting any data. 

### SeaTunnel Version

2.3.5

### SeaTunnel Config

```conf
none
```

### Running Command

```shell
none
```

### Error Exception

```log
none
```

### Zeta or Flink or Spark Version

flink 1.15.4

### Java or Scala Version

java 1.8

### Screenshots

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by litiliu**
### Purpose of this pull request

this close #7230 

### How was this patch tested?

 
new UT case has been added.

### Check list

* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [ ] Update the [`release-note`](https://github.com/apache/seatunnel/blob/dev/release-note.md).

### CODE REVIEW

File: seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/source/KafkaSourceSplitEnumerator.java [hunk 1]
```diff
@@ -194,19 +195,26 @@ public void addSplitsBack(List<KafkaSourceSplit> splits, int subtaskId) {
...
     }
 
+    @VisibleForTesting
+    protected static Map<TopicPartition, KafkaSourceSplit> resetOffset(
+            List<KafkaSourceSplit> splits, Map<TopicPartition, Long> listOffsets) {
+        splits.forEach(
+                split -> {
+                    split.setStartOffset(split.getEndOffset() + 1);
+                    split.setEndOffset(listOffsets.get(split.getTopicPartition()));
+                    split.setRecover(true);
```
Comment (Hisoka-X) [inline seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/source/KafkaSourceSplitEnumerator.java]: why not just remove current `split` from `assignedSplit` map? Then we can re-assgin it normally.

**Review by Hisoka-X — COMMENTED** [review]
Thanks @litiliu for created this PR! I have a question.

File: seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/source/KafkaSourceSplitEnumerator.java [hunk 2]
```diff
@@ -180,10 +181,20 @@ public void close() throws IOException {
     @Override
     public void addSplitsBack(List<KafkaSourceSplit> splits, int subtaskId) {
         if (!splits.isEmpty()) {
-            pendingSplit.putAll(convertToNextSplit(splits));
+            Map<TopicPartition, ? extends KafkaSourceSplit> nextSplit = convertToNextSplit(splits);
+            addSplitsBack(nextSplit, pendingSplit, assignedSplit);
```
Comment (Hisoka-X) [inline seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/source/KafkaSourceSplitEnumerator.java]: I think the new `addSplitsBack` is unnecessary. Just put logic in here and not create new method.

File: seatunnel-connectors-v2/connector-kafka/src/test/java/org/apache/seatunnel/connectors/seatunnel/kafka/source/KafkaSourceSplitEnumeratorTest.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+import java.util.HashMap;
+import java.util.Map;
+
+class KafkaSourceSplitEnumeratorTest {
+
+    @Test
+    void addSplitsBack() {
```
Comment (Hisoka-X) [inline seatunnel-connectors-v2/connector-kafka/src/test/java/org/apache/seatunnel/connectors/seatunnel/kafka/source/KafkaSourceSplitEnumeratorTest.java:39]: This test case can be optimized. For example, we can directly construct a new `KafkaSourceSplitEnumerator` class, and then initialize our `pendingSplit` and `assginSplit` by override the `open` method. Then we can call `public void addSplitsBack(List<KafkaSourceSplit> splits, int subtaskId)` for testing. Please refer https://github.com/apache/seatunnel/blob/1e7c78dd53f6fd1e75806fc5f4fb3b26676d1f88/seatunnel-engine/seatunnel-engine-server/src/test/java/org/apache/seatunnel/engine/server/resourcemanager/FakeResourceManager.java#L43

**Review by Hisoka-X — COMMENTED** [review]
Thanks @litiliu for update!

**Review by Hisoka-X — APPROVED** [review]

**Review by hailin0 — APPROVED** [review]

**Pull request merged**